### PR TITLE
Added support for non-uniform scaling and padding of shapes.

### LINKED ISSUE #29: Support nonuniform shape scaling.
Now it is only possible to use a single number as the shape's scale. But I create shapes read from a URDF as meshes, where nonuniform scaling may occur. I'd like to be able to propagate this nonuniform scaling into the resulting shapes::Shape, too.

### PR DESCRIPTION
**Pull request by peci1**
Resolves #29 .

### CODE REVIEW

**Review by rhaschke — COMMENTED** [review]
Hm, I'm not yet convinced of this feature: scaling and padding are intended for collision checking. 
Why is it beneficial to scale or pad differently along different axis dimensions, e.g. stretch a cylinder?

As pointed out in https://github.com/ros-planning/geometric_shapes/issues/29#issue-64066582, you are concerned about non-uniformly scaled meshes when they are loaded from URDF. Aren't those meshes scaled before they are associated with a mesh shape? 
Hence scaling uniformly afterwards, should keep the aspect ratio requested in the URDF, shouldn't it?

Conceptually, I don't like the approach of using a common virtual methods `scale(·,·,·)` and `padd(·,·,·)` for all shapes, although these method are _not_ meaningful in its generality to most of the shapes. This is different to `scale(·)` and `padd(·)`, which are meaningful to _all_ of the shapes (except planes).
As developers, who want to use these new functions, need to type-switch anyways to handle these specific cases, I would suggest to use type-specific methods instead (if desired at all).

**Comment by peci1** [comment]
Ahh, right, I haven't been aware of `shapes::createMeshFromResource(path, scale)`. This way meshes can really be pre-scaled and thus non-uniform scaling of the shape is not required.

However, nobody says this library can only be used for shape masks. Maybe somebody will want to use it elsewhere and would make use of the non-uniform scaling options. 

If you look at the code, you'll see that scaling that would make the shapes non-primitive (like making an "ellipsoid cylinder") are not supported. But I still think it's good to be able to scale (or more importantly pad) e.g. radius differently than height. Of course you can do the computations before constructing the shapes, but as seen in e.g. #107, even such simple computations can be written wrong. That's why I think it's better the library can do it natively.

**Comment by rhaschke** [comment]
> However, nobody says this library can only be used for shape masks.

This is correct. Indeed the feature might be useful in other contexts. 
On the other side, the library is primarily a support lib for MoveIt's collision checking. As such, I consider this feature as a kind of code bloat, which will require extra maintenance effort in the future :wink:
However, I'm fine to integrate this feature in general.

> If you look at the code, you'll see that scaling that would make the shapes non-primitive (like making an "ellipsoid cylinder") are not supported.

I have seen this and I exactly don't like these runtime checks. As discussed above, callers of the new methods need to handle different shapes separately anyway, because e.g. `Sphere::scale(1,2,3)` is an invalid call and will just result in an error message on the console. For this reason, I always prefer an API that forbids invalid usage. Hence, my suggestion is to use class-specific methods, e.g. `Cylinder::scale(radius, length)`.

**Comment by rhaschke** [comment]
@peci1, most of your pending commits have merge conflicts. Could you please rebase them onto latest `melodic-devel`? Thanks.

**Comment by peci1** [comment]
Hi, I reimplemented this as you suggested - i.e. each shape only allows scaling the dimensions it really has.

File: src/shapes.cpp [hunk 1]
```diff
@@ -298,40 +298,84 @@ void Sphere::scaleAndPadd(double scale, double padding)
...
+void Cylinder::scale(double scaleRadius, double scaleLength)
 {
-  const auto tmpRadius = radius * scale + padding;
-  const auto tmpLength = length * scale + 2.0 * padding;
+  scaleAndPadd(scaleRadius, scaleLength, 0.0, 0.0);
+}
+
+void Cylinder::padd(double paddRadius, double paddLength)
+{
+  scaleAndPadd(0.0, 0.0, paddRadius, paddLength);
```
Comment (v4hn) [inline src/shapes.cpp]:
```suggestion
  scaleAndPadd(1.0, 1.0, paddRadius, paddLength);
```

I might be completely mistaken, but why did were these not caught by your tests? :confused:

File: src/shapes.cpp [hunk 2]
```diff
@@ -298,40 +298,84 @@ void Sphere::scaleAndPadd(double scale, double padding)
...
 
-void Box::scaleAndPadd(double scale, double padding)
+void Cone::scale(double scaleRadius, double scaleLength)
+{
+  scaleAndPadd(scaleRadius, scaleLength, 0.0, 0.0);
+}
+
+void Cone::padd(double paddRadius, double paddLength)
+{
+  scaleAndPadd(0.0, 0.0, paddRadius, paddLength);
```
Comment (v4hn) [inline src/shapes.cpp]:
```suggestion
  scaleAndPadd(1.0, 1.0, paddRadius, paddLength);
```

File: src/shapes.cpp [hunk 3]
```diff
@@ -298,40 +298,84 @@ void Sphere::scaleAndPadd(double scale, double padding)
...
 
-void Mesh::scaleAndPadd(double scale, double padding)
+void Box::scale(double scaleX, double scaleY, double scaleZ)
+{
+  scaleAndPadd(scaleX, scaleY, scaleZ, 0.0, 0.0, 0.0);
+}
+
+void Box::padd(double paddX, double paddY, double paddZ)
+{
+  scaleAndPadd(0.0, 0.0, 0.0, paddX, paddY, paddZ);
```
Comment (v4hn) [inline src/shapes.cpp]:
```suggestion
  scaleAndPadd(1.0, 1.0, 1.0, paddX, paddY, paddZ);
```

File: src/shapes.cpp [hunk 4]
```diff
@@ -360,23 +404,37 @@ void Mesh::scaleAndPadd(double scale, double padding)
...
 }
 
+void Mesh::scale(double scaleX, double scaleY, double scaleZ)
+{
+  scaleAndPadd(scaleX, scaleY, scaleZ, 0.0, 0.0, 0.0);
+}
+
+void Mesh::padd(double paddX, double paddY, double paddZ)
+{
+  scaleAndPadd(0.0, 0.0, 0.0, paddX, paddY, paddZ);
```
Comment (v4hn) [inline src/shapes.cpp]:
```suggestion
  scaleAndPadd(1.0, 1.0, 1.0, paddX, paddY, paddZ);
```

**Review by v4hn — COMMENTED** [review]
I approve after the inline comments have been addressed (or refuted).

**Comment by peci1** [comment]
Really good catch, @v4hn. The tests did not cover this because I only tested the nonuniform `scaleAndPadd`, and not the nonuniform `scale` and `padd` functions. I added tests which fail before the suggested fixes.

**Review by v4hn — APPROVED** [review]

**Review by rhaschke — APPROVED** [review]

**Pull request merged**